### ISSUE TITLE
Compiler crash: failed requirement in `JSEncoding.encodeRTStringMethodSym`

### ISSUE THREAD

**Issue by sarahgerweck** [scala-js/scala-js#3211]
When compiling [simple code that contains a structural type](https://github.com/sarahgerweck/scalajs-failure/blob/master/Fail.scala) with ScalaJS 0.6.21, I get a stack trace with the following top frames:

```
[error] scala.tools.nsc.Global.require(Global.scala:279)
[error] org.scalajs.core.compiler.JSEncoding.encodeRTStringMethodSym(JSEncoding.scala:159)
[error] org.scalajs.core.compiler.JSEncoding.encodeRTStringMethodSym$(JSEncoding.scala:156)
[error] org.scalajs.core.compiler.GenJSCode.encodeRTStringMethodSym(GenJSCode.scala:31)
[error] org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genApplyDynamic$7(GenJSCode.scala:3770)
[error] org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genApplyDynamic$7$adapted(GenJSCode.scala:3754)
[error] scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:789)
[error] scala.collection.immutable.List.foreach(List.scala:389)
[error] scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:788)
```

I've put a GitHub repo at https://github.com/sarahgerweck/scalajs-failure that can be cloned, which will show the failure on `sbt compile`.

I'm not clear on whether ScalaJS is really intended to support the kind of structural type dispatch I'm doing here. Even if not, it would be nice to have an error message to that effect instead of a blanket failure.

Based on some Gitter conversation, it seems like there may be some confusion with the ScalaJS compiler trying to handle the `.chars` method on a `String`, which of course is one thing that could be passed into this method (though not something I ever actually to pass in my real code).

**Comment by sjrd** [comment]
Thanks for the report. It should be supported.

Although it would fail at runtime if actually called with a `String`--because `String.chars` is not implemented--, it shouldn't prevent your code from working if called with something that actually had implemented a `chars` method.

**Comment by sjrd** [comment]
Completed with full reproduction, and simplified to
```scala
object LexerUtil {
  def reflectiveLongest[A <: { def chars: String }](data: A): Unit = {
    println(data.chars)
  }
}

class Foo {
  def chars: String = "the chars"
}

object HelloWorld {
  def main(args: Array[String]): Unit = {
    LexerUtil.reflectiveLongest(new Foo)
  }
}
```
So, one piece of good news: as I suspected, this bug does not happen in Scala.js 1.0.0-M2 anymore. However, although there is no compiler crash, we still get a spurious linking error:
```
[error] Referring to non-existent method java.lang.CharSequence.chars()java.util.stream.IntStream
[error]   called from scala.collection.mutable.StringBuilder.chars()java.util.stream.IntStream
[error]   called from scala.collection.mutable.StringBuilder.chars()
[error]   called from helloworld.LexerUtil$.reflectiveLongest(java.lang.Object)void
[error]   called from helloworld.HelloWorld$.main([java.lang.String)void
[error]   called from core module module initializers
[error] involving instantiated classes:
[error]   scala.collection.mutable.StringBuilder
[error]   helloworld.LexerUtil$
[error]   helloworld.HelloWorld$
```
I believe we can easily "fix" 0.6.x to have the linking error instead of the compiler crash as well (simply by removing the offending `require`), but that's still not good enough anyway.

The reason for the linking error is that the Scala.js compiler generates the following IR for the call in `reflectiveLongest`:
```scala
  def reflectiveLongest__O__V(data: any) {
    mod:s_Predef$.println__O__V({
      val x: any = data;
      if (x.isInstanceOf[jl_String]) {
        x.asInstanceOf[jl_String].chars__ju_stream_IntStream()
      } else {
        x.chars__()
      }
    })
  }
```
In which we see a full *non-reflective* call to `String.chars(): IntStream`, which is not implemented.

If we called `String.chars()` directly in the source code, that linking error would be justified, but in this case it is initially called reflectively, so there is no reason that it should fail to link. It should instead fail at run-time with a `TypeError`.

I am not yet sure how to best fix this at the moment. The proper, long-term fix would be to add complete support for reflective proxies in hijacked classes, but that's definitely not possible in 0.6.x, and not that easy in 1.x either.

**Comment by sjrd** [comment]
Hum there's a bigger problem. Even if I fix that linking error, there is another one happening through `scala.collection.mutable.StringBuilder`, and through the bridge that scalac automatically generate for the default method `CharSequence.chars()`, and which looks like:
```scala
  def chars__ju_stream_IntStream(): ju_stream_IntStream = {
    this.jl_CharSequence::chars__ju_stream_IntStream()
  }
```
So in fact that method (correctly) becomes reachable through the reflective call `x.chars__()`, which then tries to reach `CharSequence.chars()` which does not exist. The linking error is:
```
[error] Referring to non-existent method java.lang.CharSequence.chars()java.util.stream.IntStream
[error]   called from scala.collection.mutable.StringBuilder.chars()java.util.stream.IntStream
[error]   called from scala.collection.mutable.StringBuilder.chars()
[error]   called from helloworld.LexerUtil$.reflectiveLongest(java.lang.Object)void
[error]   called from helloworld.HelloWorld$.main([java.lang.String)void
[error]   called from core module module initializers
[error] involving instantiated classes:
[error]   scala.collection.mutable.StringBuilder
[error]   helloworld.LexerUtil$
[error]   helloworld.HelloWorld$
```
I do not think that *that* linking error is avoidable in any reasonable way 😖

**Comment by sjrd** [comment]
The core issue is also not related to `chars()` nor `String` nor `CharSequence` nor default methods. It is reproducible with a custom class hierarchy and custom method:
```scala
package helloworld

import language.reflectiveCalls

class WellBehaved {
  def problematicMethod(): Int = 5
}

class IllBehaved {
  def problematicMethod(): Unit = Thread.sleep(50L)
}

object HelloWorld {
  def main(args: Array[String]): Unit = {
    val wellBehaved = new WellBehaved
    problem(wellBehaved)

    new IllBehaved // the very presence of this line causes the issue
  }

  def problem(x: Any { def problematicMethod(): Int }): Unit =
    println(x.problematicMethod())
}
```
Linking error:
```
[error] Referring to non-existent method java.lang.Thread$.sleep(long)void
[error]   called from helloworld.IllBehaved.problematicMethod()void
[error]   called from helloworld.IllBehaved.problematicMethod()
[error]   called from helloworld.HelloWorld$.problem(java.lang.Object)void
[error]   called from helloworld.HelloWorld$.main([java.lang.String)void
[error]   called from core module module initializers
[error] involving instantiated classes:
[error]   helloworld.IllBehaved
[error]   helloworld.HelloWorld$
```
That even kills hopes of fixing the original snippet of code by implementing `CharSequence.chars()` (which would require implementing `Stream`s!) or any other one-off solution. And at the same time, in the general case, the linking error is actually *correct*.

At this point, I'm afraid this is simply unfixable.

@gzm0 Any other idea?

**Comment by sarahgerweck** [comment]
@sjrd Is this a fair characterization of the issue?

If you have a method that contains a structural type, ScalaJS has to be able to successfully compile any methods that have the right name (and parameters?) so that it can create link points in case they're used for structural dispatch? (In my `chars` case, because it requires streams. In the `IllBehaved` case, because of the `Thread.sleep`.)


In both my use case and yours, the return type of the method (either `CharSequence.chars` or `IllBehaved.problematicMethod`) disqualifies those objects from satisfying the structural type in question. In my example, `CharSequences.chars()` returns an `IntSequence`, which does not fulfill the required `A <: { def chars: String }`. The same goes for `Unit` vs `Int` in your example. Is this part of the problem, or would checking for compatible return types be a way to reduce the number of methods that have to be linkable?

**Comment by sjrd** [comment]
It's a pretty good characterization, yes. Parameters are taking into account in addition to the method name, yes.

Checking the result type is not possible, as it would disqualify *valid* use cases (because of covariant result types). We initially tried that, a long time ago, and pretty much nothing worked. Anyway, even if it did *reduce* the problematic cases, it would not solve them.

**Comment by gzm0** [comment]
Yes, I agree. We cannot fix this without additional analysis. In your snippet, the linker would have to prove that `problem` is never passed a `IllBehaved`. This is simply unrealistic in the general case.

An option would be to introduce structural type tracking in the compiler where it tracks what named types are upcast into what structural types. Then this could be avoided, but it is probably very difficult to write :-/

**Comment by sjrd** [comment]
The compiler crash was fixed in 7abedec0d3c0dff3eedba82d333cb7be0c59d1ed.

The rest of the issue with the linking errors, however, I'm afraid I have to close as a won't fix.
